### PR TITLE
Improvements of releases list and tags list

### PR DESCRIPTION
**Pull request by Zettat123**
Follow #23465 and #25624

This PR introduces the following improvements:
- We do not need to call `GetTags` to get tags because tags have been loaded by `RepoAssignment`
https://github.com/go-gitea/gitea/blob/ef90fdbd1d7e1f62ed5bf18757e00e97817eb602/modules/context/repo.go#L663-L668
- Similarly, the number of tags and releases also have been loaded by `RepoAssignment`, so the related code has been removed from the handlers. The query condition of `GetReleaseCountByRepoID` in `RepoAssignment` has been changed to include draft releases.
https://github.com/go-gitea/gitea/blob/ef90fdbd1d7e1f62ed5bf18757e00e97817eb602/modules/context/repo.go#L538-L551
- `releasesOrTags` function has been removed. The code for rendering releases list and tags list moved to `Releases` and `TagList` respectively.

### CODE REVIEW

**Review by wolfogre — APPROVED** [review]

File: routers/web/repo/release_test.go [hunk 1]
```diff
@@ -73,6 +74,12 @@ func TestNewReleasesList(t *testing.T) {
 	test.LoadGitRepo(t, ctx)
 	t.Cleanup(func() { ctx.Repo.GitRepo.Close() })
 
+	var err error
+	ctx.Data["NumReleases"], err = repo_model.GetReleaseCountByRepoID(ctx, ctx.Repo.Repository.ID, repo_model.FindReleasesOptions{
+		IncludeDrafts: ctx.Repo.CanWrite(unit_model.TypeReleases),
+	})
+	assert.NoError(t, err)
```
Comment (wxiaoguang) [inline routers/web/repo/release_test.go]: Is there an assertion for `ctx.Data["NumReleases"]`?

Comment (wxiaoguang) [reply routers/web/repo/release_test.go]: Hmm .... I think the test code is not ideal. Calling `Releases` directly makes the middlewares never executed.


If it doesn't test `NumReleases` logic, it's better to remove this code block.

Comment (Zettat123) [reply routers/web/repo/release_test.go]: Sorry for replying so late. I agree with your opinion that the test code is not ideal. Right now this unit test cannot work without the `RepoAssignment` middleware. I tried moving it to integration tests, then I found there was already a `TestViewReleaseListNoLogin` which tests the release list.

https://github.com/go-gitea/gitea/blob/5600504eed2473691b86962be2f9733e6bfce879/tests/integration/release_test.go#L134-L174

Maybe we can remove `TestNewReleasesList` from unit tests?

**Review by wxiaoguang — COMMENTED** [review]
Why `TestNewReleasesList` is removed? Could it be kept?

**Comment by Zettat123** [comment]
> Why `TestNewReleasesList` is removed? Could it be kept?

Because `ctx.Data["NumReleases"]` is required in `TestNewReleasesList`. However, `NumReleases` is assigned in the  `RepoAssignment` middleware and this test calls the handler(`Release`) directly without using the middleware. So `TestNewReleasesList` cannot work now. 

In addition, as my previous comment said, I think the `TestViewReleaseListNoLogin` integration test is able to test the logic of the handler.

**Comment by wxiaoguang** [comment]
The test code and `TargetBehind` was introduced by #24470

If I understand correctly, only the removed code tested the `TargetBehind` ?

**Comment by Zettat123** [comment]
> The test code and `TargetBehind` was introduced by #24470
> 
> If I understand correctly, only the removed code tested the `TargetBehind` ?

To test `TargetBehind`, I added `TestNewReleasesList` back and renamed it to `TestCalReleaseNumCommitsBehind` in a8050b7cc714aff9088fc784f4a8bb480374f78e

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**